### PR TITLE
[JBPM-9291] AMQ Stream (Kafka) Integration WIH / workflow component

### PR DESCRIPTION
**Pull request by elguardian**
upstream jira
https://issues.redhat.com/browse/JBPM-9291

### CODE REVIEW

**Comment by elguardian** [comment]
depends on https://github.com/kiegroup/droolsjbpm-build-bootstrap/pull/1434

**Comment by tsurdilo** [comment]
this should be in https://github.com/kiegroup/jbpm-work-items and not in jbpm imo.

File: jbpm-workitems/jbpm-workitems-kafka/.gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+/target/
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-workitems/jbpm-workitems-kafka/.gitignore:1]: done

File: jbpm-workitems/jbpm-workitems-kafka/src/main/java/org/jbpm/workitems/kafka/KafkaSendTaskWorkItemHandler.java [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+import org.slf4j.LoggerFactory;
+
+import static java.util.Collections.emptyMap;
+
+@Wid(widfile = "KafkaSendTaskDefinitions.wid", name = "KafkaSendTask",
+     displayName = "KafkaSendTask",
+     defaultHandler = "mvel: new org.jbpm.workitems.kafka.KafkaSendTaskWorkItemHandler()",
+     documentation = "${artifactId}/index.html",
+     category = "${artifactId}",
+     icon = "KafkaSendTask.png",
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-workitems/jbpm-workitems-kafka/src/main/java/org/jbpm/workitems/kafka/KafkaSendTaskWorkItemHandler.java:57]: done

[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-workitems/jbpm-workitems-kafka/src/main/java/org/jbpm/workitems/kafka/KafkaSendTaskWorkItemHandler.java:57]: done

File: jbpm-workitems/jbpm-workitems-kafka/src/main/java/org/jbpm/workitems/kafka/KafkaSendTaskWorkItemHandler.java [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+        }
+        try {
+            ByteArrayOutputStream bout = new ByteArrayOutputStream();
+            ObjectOutputStream oout = new ObjectOutputStream(bout);
+            oout.writeObject(data);
+
+            return bout.toByteArray();
+        } catch (IOException e) {
+            throw new RuntimeException("Error during serialization for kafka server", e);
+        }
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-workitems/jbpm-workitems-kafka/src/main/java/org/jbpm/workitems/kafka/KafkaSendTaskWorkItemHandler.java:202]: done

File: jbpm-workitems/jbpm-workitems-kafka/src/main/java/org/jbpm/workitems/kafka/KafkaSendTaskWorkItemHandler.java [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+
+
+    @Override
+    public void abortWorkItem(WorkItem workItem, WorkItemManager manager) {
+        // do nothing
+    }
+
+    @Override
+    public void close() {
+        kafkaProducer.close();
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-workitems/jbpm-workitems-kafka/src/main/java/org/jbpm/workitems/kafka/KafkaSendTaskWorkItemHandler.java:181]: done

File: jbpm-workitems/jbpm-workitems-kafka/src/main/java/org/jbpm/workitems/kafka/KafkaSendTaskWorkItemHandler.java [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+package org.jbpm.workitems.kafka;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.ObjectOutputStream;
+import java.util.Collections;
+import java.util.Map;
+import java.util.HashMap;
+import java.util.Properties;
+import java.util.concurrent.ConcurrentHashMap;
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-workitems/jbpm-workitems-kafka/src/main/java/org/jbpm/workitems/kafka/KafkaSendTaskWorkItemHandler.java:25]: done

File: jbpm-workitems/jbpm-workitems-kafka/src/main/java/org/jbpm/workitems/kafka/KafkaSendTaskWorkItemHandler.java [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+import java.util.concurrent.Future;
+
+import org.apache.kafka.clients.producer.Callback;
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-workitems/jbpm-workitems-kafka/src/main/java/org/jbpm/workitems/kafka/KafkaSendTaskWorkItemHandler.java:28]: done

File: jbpm-workitems/jbpm-workitems-kafka/src/main/java/org/jbpm/workitems/kafka/KafkaSendTaskWorkItemHandler.java [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+    public void abortWorkItem(WorkItem workItem, WorkItemManager manager) {
+        // do nothing
+    }
+
+    @Override
+    public void close() {
+        kafkaProducer.close();
+    }
+
+    protected void addPropertyIfExists(String propertyName, Object properyValue, ProducerRecord<String, byte[]> record) {
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-workitems/jbpm-workitems-kafka/src/main/java/org/jbpm/workitems/kafka/KafkaSendTaskWorkItemHandler.java:184]: done

File: jbpm-workitems/jbpm-workitems-kafka/pom.xml [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+  <modelVersion>4.0.0</modelVersion>
+  <parent>
+    <groupId>org.jbpm</groupId>
+    <artifactId>jbpm-workitems</artifactId>
+    <version>7.42.0-SNAPSHOT</version>
+  </parent>
+
+  <artifactId>jbpm-workitems-kafka</artifactId>
+
+  <name>REST</name>
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-workitems/jbpm-workitems-kafka/pom.xml:15]: done

File: jbpm-workitems/jbpm-workitems-kafka/pom.xml [hunk 9]
```diff
@@ -0,0 +1,238 @@
+<?xml version="1.0"?>
+<project
+  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd"
+  xmlns="http://maven.apache.org/POM/4.0.0"
+  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
+  <modelVersion>4.0.0</modelVersion>
+  <parent>
+    <groupId>org.jbpm</groupId>
+    <artifactId>jbpm-workitems</artifactId>
+    <version>7.42.0-SNAPSHOT</version>
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-workitems/jbpm-workitems-kafka/pom.xml:10]: done

File: jbpm-workitems/jbpm-workitems-kafka/src/main/java/org/jbpm/workitems/kafka/KafkaSendTaskWorkItemHandler.java [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+        addPropertyIfExists("KIE_WorkItemId", workItem.getId(), record);
+
+        Future<RecordMetadata> workItemUnitOfWork = kafkaProducer.send(record);
+        try {
+            RecordMetadata metadata = workItemUnitOfWork.get();
+            Map<String, Object> results = new HashMap<>();
+            results.put("kafka_metadata", metadata);
+            manager.completeWorkItem(workItem.getId(), results);
+        } catch (Exception e) {
+            handleException(e);
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-workitems/jbpm-workitems-kafka/src/main/java/org/jbpm/workitems/kafka/KafkaSendTaskWorkItemHandler.java:166]: there is already a logic for handling the error. logic is not needed as the handle error will do that for us.

**Comment by elguardian** [comment]
Jenkins retest this please.

**Comment by elguardian** [comment]
@tsurdilo we are providing OOTB of this component. so it goes with the same strategy as jms, email, and others.
The question I dealt with Kris is about removing it from that wih repo or not. wdyt ?

**Comment by elguardian** [comment]
Jenkins retest this please.

**Comment by elguardian** [comment]
Jenkins retest this please.

**Review by afalhambra — APPROVED** [review]
Nice work!

**Pull request closed without merging**